File: Leetcode/target_sum.py
# https://leetcode.com/problems/target-sum/
'''
    Author - Subhajit Das
    University of Engineering and Management, Kolkata
    17/10/2019
'''
from sys import maxsize

class Solution(object):
    def findTargetSumWays(self, nums, S):
        """
        :type nums: List[int]
        :type S: int
        :rtype: int
        """
        self.count = 0
        self.targetSum = S
        self.mem = [[maxsize]*2001 for _ in nums]

        return self.findTargetHelper(nums, 0, 0)

    def findTargetHelper(self, nums, idx, sumTillNow):
        if(idx == len(nums)):
            return  1 if(sumTillNow == self.targetSum) else 0
        
        if(self.mem[idx][sumTillNow+1000] != maxsize):
            return self.mem[idx][sumTillNow+1000]
        
        self.mem[idx][sumTillNow+1000] = self.findTargetHelper(nums, idx+1, sumTillNow + nums[idx]) + self.findTargetHelper(nums, idx+1, sumTillNow - nums[idx])
        
        return self.mem[idx][sumTillNow+1000]

File: Leetcode/TargetSum.java
// https://leetcode.com/problems/target-sum/
/*
    Author @Subhajit Das (sdsubhajitdas.github.io)
    SWE @Turbot HQ India PVT Ltd.
    19/06/2021
*/

import java.util.*;


class Solution {
    public int findTargetSumWays(int[] nums, int target) {
        
        int totalSum = 0;
        for(int num: nums) totalSum+=num;

        if(target > totalSum || (totalSum + target)%2 == 1){
            return 0;
        }
        
        target = (target + totalSum)/2;
        int dp[][] = new int[nums.length+1][target+1];
        for(int row=0; row<=nums.length; row++) dp[row][0] = 1;
                
        
        for(int idx=1; idx<=nums.length; idx++) {
            for(int sum=0; sum<=target; sum++) {
                
                if(nums[idx-1] <= sum) {
                    dp[idx][sum] = dp[idx-1][sum] + dp[idx-1][sum-nums[idx-1]];
                }
                else {
                    dp[idx][sum] = dp[idx-1][sum];
                }
            }
        }
        
        return dp[nums.length][target];
    }
}